hashjoin.go

```diff
@@ -0,0 +1,393 @@
+	}
+}
+
+func (prober *hashJoinProber) start(flow dataFlow, eqColIdx int, outCols []int, outFlow *dataFlow) {
+	eqCol := flow.b[eqColIdx].(intColumn)[:batchRowLen]
+
+	prober.computeBucket(eqCol)
+	prober.lookupInitial()
+
+	nToCheck := len(prober.toCheck)
+
+	for nToCheck > 0 {
+		nToCheck = prober.check(eqCol)
+		prober.findNext(nToCheck)
+	}
+
+	groupId := prober.groupId
+	results := make([]int, 0)
+	resultsIdx := make([]int, 0)
+
+	for i := 0; i < flow.n; i++ {
+		if groupId[i] != 0 {
+			results = append(results, groupId[i])
+			resultsIdx = append(resultsIdx, i)
+		}
+	}
+
+	nBuildOutCols := prober.hashTable.nOutCols
+	nResults := len(results)
+
+	for j := 0; j < nBuildOutCols; j++ {
+		outCol := outFlow.b[j].(intColumn)[:batchRowLen]
+		valCol := prober.hashTable.values[j]
+
+		for i := 0; i < nResults; i++ {
+			outCol[i] = valCol[results[i]]
+		}
+	}
+
+	for j, colIdx := range outCols {
+		outCol := outFlow.b[j + nBuildOutCols].(intColumn)[:batchRowLen]
+		valCol := flow.b[colIdx].(intColumn)[:batchRowLen]
+
+		for i := 0; i < nResults; i++ {
+			outCol[i] = valCol[resultsIdx[i]]
+		}
+	}
+
+	outFlow.n = nResults
+}
+
+// hashJoinGroupBuilder is an alternative builder that pre-loads all build
+// relation batches before building the hash table.
+type hashJoinGroupBuilder struct {
+	hashTable *hashTableInt
+
+	totalSize int
+
+	// bucket stores the hashed value of every key in the build relation, and
+	// thereby contributes additional memory in comparison to the
+	// hashJoinBuilder.
+	bucket []int
+}
+
+func (hashTable *hashTableInt) growGroup(amount int) {
+	hashTable.keys = append(hashTable.keys, make(intColumn, amount)...)
+
+	for i := 0; i < hashTable.nOutCols; i++ {
+		hashTable.values[i] = append(hashTable.values[i], make(intColumn, amount)...)
+	}
+
+	hashTable.allocated += amount
+}
+
+func (hashTable *hashTableInt) growNext(amount int) {
+	hashTable.next = append(hashTable.next, make([]int, amount)...)
+}
+
+func (hashTable *hashTableInt) insertKey(hashKey int, id int) {
+	hashTable.count++
+	hashTable.next[id] = hashTable.first[hashKey]
+	hashTable.first[hashKey] = id
+}
+
+func makeHashJoinGroupBuilder(hashTable *hashTableInt) *hashJoinGroupBuilder {
+	return &hashJoinGroupBuilder{
+		hashTable: hashTable,
+	}
+}
+
+func (builder *hashJoinGroupBuilder) insertBatch(flow dataFlow, eqColIdx int, outCols []int, batchSize int) {
+	builder.hashTable.growGroup(batchSize)
+
+	eqCol := flow.b[eqColIdx].(intColumn)[:batchRowLen]
+
+	for i := 0; i < batchSize; i++ {
+		builder.hashTable.keys[i + builder.totalSize + 1] = eqCol[i]
+	}
+
+	for idx, colIdx := range outCols {
+		outCol := flow.b[colIdx].(intColumn)[:batchRowLen]
+		builder.spread(outCol, idx, batchSize)
+	}
+
+	builder.totalSize += batchSize
+}
+
+func (builder *hashJoinGroupBuilder) spread(outCol intColumn, valIdx int, batchSize int) {
+	valCol := builder.hashTable.values[valIdx]
+
+	for i:= 0; i < batchSize; i++ {
+		valCol[i + builder.totalSize + 1] = outCol[i]
```

Comment:
another + 1 - why? Seems like you could get rid of these everywhere, maybe, unless I'm missing something.